src/assisted_installer_controller/assisted_installer_controller.go

```diff
@@ -305,43 +296,84 @@ func (c controller) unallocatedMachines(bmhList metal3v1alpha1.BareMetalHostList
 	return unallocatedList, nil
 }
 
-func (c controller) updateBMHs(bmhList metal3v1alpha1.BareMetalHostList, machineList *mapiv1beta1.MachineList) bool {
+func (c controller) updateBMHAnnotations(bmh *metal3v1alpha1.BareMetalHost, provisioningExists bool) (bool, error) {
+	annotations := bmh.GetAnnotations()
+	content := []byte(annotations[metal3v1alpha1.StatusAnnotation])
+
+	if provisioningExists {
+		// when baremetal operator is enabled, we need to remove the pausedAnnotation
+		// to indicated that the statusAnnotation is available. This is only on
+		// the master nodes.
+		if _, ok := annotations[metal3v1alpha1.PausedAnnotation]; ok {
+			c.log.Infof("Removing pausedAnnotation from BMH host %s", bmh.Name)
+			delete(annotations, metal3v1alpha1.PausedAnnotation)
+			return true, nil
+		}
+		return false, nil
+	}
+
+	if annotations[metal3v1alpha1.StatusAnnotation] == "" {
+		c.log.Infof("Skipping setting status of BMH host %s, status annotation not present", bmh.Name)
+		return false, nil
+	}
+	objStatus, err := c.unmarshalStatusAnnotation(content)
+	if err != nil {
+		c.log.WithError(err).Errorf("Failed to unmarshal status annotation of %s", bmh.Name)
+		return false, err
+	}
+	bmh.Status = *objStatus
+	if bmh.Status.LastUpdated.IsZero() {
+		// Ensure the LastUpdated timestamp in set to avoid
+		// infinite loops if the annotation only contained
+		// part of the status information.
+		t := metav1.Now()
+		bmh.Status.LastUpdated = &t
+	}
+	err = c.kc.UpdateBMHStatus(bmh)
+	if err != nil {
+		c.log.WithError(err).Errorf("Failed to update status of BMH %s", bmh.Name)
+		return false, err
+	}
+	delete(annotations, metal3v1alpha1.StatusAnnotation)
+	return true, nil
```

Comment:
Updating the Status changes the version of the resource, so an immediate write to change the annotations will always fail (it'll work the next time with a fresh copy of the resource though). This appears to be the case here, although it pre-dates this patch.